### PR TITLE
HW6 Solution

### PR DESCRIPTION
**Pull request by dsergio**

### CODE REVIEW

File: Assignment6/Array.cs [hunk 1]
```diff
@@ -1,10 +1,84 @@
 ﻿using System;
+using System.Collections;
 using System.Collections.Generic;
 
 namespace Assignment6
 {
-    public class Array<T>
+    public class Array<T> : System.Collections.Generic.ICollection<T>
```
Comment (RSXEnthusiast) [inline Assignment6/Array.cs]:
```suggestion
    public class Array<T> : ICollection<T>
```

You don't need the whole path here.

File: Assignment6/Array.cs [hunk 2]
```diff
@@ -1,10 +1,84 @@
...
 using System.Collections.Generic;
 
 namespace Assignment6
 {
-    public class Array<T>
+    public class Array<T> : System.Collections.Generic.ICollection<T>
     {
-        
+        private Dictionary<int, T> Data { get; set; }
+        public int Size { get; private set; }
```
Comment (RSXEnthusiast) [inline Assignment6/Array.cs]:
```suggestion
        public int Capacity { get; private set; }
```

He did specifically say to make a Capacity property in the assignment. Not sure the naming would matter, but I'd change it.

File: ArrayTests/GlobalSuppressions.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
+﻿// This file is used by Code Analysis to maintain SuppressMessage
+// attributes that are applied to this project.
+// Project-level suppressions either have no target or are given
+// a specific target and scoped to a namespace, type, member, etc.
+
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Naming", "CA1707:Identifiers should not contain underscores", Justification = "Test Methods Should Contain Underscores", Scope = "member", Target = "~M:Assignment6.ArrayTests.ArrayString_SetGetInBounds_ReturnsCorrectValue")]
```
Comment (AaronSauther) [inline ArrayTests/GlobalSuppressions.cs]: If you get rid of the target and scope tags the suppression will work for the entire ArrayTests project, I know how annoying it is to individually suppress all of those!
```suggestion
[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Naming", "CA1707:Identifiers should not contain underscores", Justification = "Test Methods Should Contain Underscores")]
```

File: Assignment6/ArrayCollection.cs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+namespace Assignment6
+{
+    /// <summary>
+    /// Extra Credit: cast operator to a C# array or a `System.Collections.Generic.List<T>`
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// 
+    public class ArrayCollection<T> : ICollection<T>
+    {
+        private List<T> Data { get; set; }
```
Comment (MarkMichaelis) [inline Assignment6/ArrayCollection.cs]: Why the setter?  Shouldn't a read-only property suffice?

File: Assignment6/ArrayCollection.cs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        public int Capacity { get; private set; }
+        public int Count => Data.Count;
+        public bool IsReadOnly { get; set; }
```
Comment (MarkMichaelis) [inline Assignment6/ArrayCollection.cs]: Should this really be a public setter?  In fact, I am somewhat surprised that there is a setter.

File: Assignment6/ArrayCollection.cs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+            }
+        }
+
+        public void CopyTo(T[] array, int arrayIndex)
+        {
+            Data.CopyTo(array, arrayIndex);
+        }
+
+        public bool Remove(T item)
+        {
```
Comment (MarkMichaelis) [inline Assignment6/ArrayCollection.cs]: What are you doing that `Data.Remove(item)` doesn't have built in functionality for?

File: Assignment6/ArrayCollection.cs [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+            {
+                return Data.Remove(item);
+            } else
+            {
+                return false;
+            }
+            
+        }
+
+        public IEnumerator<T> GetEnumerator()
```
Comment (MarkMichaelis) [inline Assignment6/ArrayCollection.cs]: Consider using an expression bodied member here:
```
public IEnumerator<T> GetEnumerator() => Data.GetEnumerator();
```

File: Assignment6/ArrayCollection.cs [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+        /// 
+        /// </summary>
+        /// <param name="arr"></param>
+        public static explicit operator ArrayCollection<T>(Array arr)
+        {
+            if (arr is null)
+            {
+                throw new ArgumentNullException(nameof(arr));
+            }
+            ArrayCollection<T> ret = new ArrayCollection<T>(arr.Length);
```
Comment (MarkMichaelis) [inline Assignment6/ArrayCollection.cs]: Can't you call the `ArrayCollection(ICollection<T> collection)` constructor?

File: ArrayTests/ArrayTests.cs [hunk 9]
```diff
@@ -0,0 +1,538 @@
...
+
+            StringBuilder sb = new StringBuilder();
+            foreach (string s in arr)
+            {
+                Assert.IsTrue(s is string);
+                sb.Append(s);
+            }
+
+            // Assert
+            Assert.AreEqual<string>("zero_one_two_three_four_five_", sb.ToString());
```
Comment (MarkMichaelis) [inline ArrayTests/ArrayTests.cs]: I recommend using the CollectionAssert class instead of the ToString() approach.

**Comment by MarkMichaelis** [comment]
##  (dsergio)

## Assignment6

| Build/Compile (Build Succeeded: 0 Error(s), 0 Warning(s))| **10** |
| --------------------- | ------ |
| Assignment compiles even though major pieces might be missing, or there are significant problems: **✔** |  **5** |
| The assignment contains only a few required elements or has significant problems. | **5** |


| Assignment  | **20** |
| --------------------- | ------ |
| Supports collecting a fixed number of items once instantiated. | ✔ |
| Has a constructor via which you can specify fixed width.  (Other constructors are welcome.) | ✔ |
| Handle null appropriately when items are missing (this is likely a judgement call so justify your answer in the comments.) | ✔ |
| Throw an exception if an item doesn't exist. | ✔ |
| Include an index operator. | ✔ |
| Support using foreach over the items in the `Array<T>` class. | ✔ |
| Support a collection initializer. | ✔ |
| Implement `System.Collections.Generic.ICollection<T>`. | ✔ |
| Ignore support for multi-dimensional arrays. | ✔ |
| Add a `Capacity` property to return the capacity (the value specified in the constructor) of the collection. | ✔ |


| Project Analysis |   |
| -----------------| - |
| Nullable Enabled: | **✔** |
| CodeReview Performed | **✔** |
| Add the code analysis packages and appropriately handle all warnings* |  **✔** |
*<sub><sup>`IntelliTectAnalyzer.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.CodeQuality.Analyzers.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.NetFramework.Analyzers.dll`. (It is okay to appropriately disable rules such as  localization/culture rules.)</sub></sup>


| ProjectName | Nullable | Solutions | Analyzers | Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| ArrayTests | Enable | Assignment6.sln |  | CS1701, CS1702 | 4 |
| Assignment6| Enable | Assignment6.sln |  | CS1701, CS1702 | 4 |


| Testing | **20** |
| --------------------- | ------ |
| Test Run Succeeded: Total tests=43, Passed=43, Failed=0, Skipped=0 | **96%** |
| Unit test project fails to compile, most test fails, or contains significant problems. | **10** |
| Most assigned work is tested or contain some errors. | **5** |
| All assigned work is tested | **5** |


| Extra Credit | **5** |
| --------------------- | ------ |
| Write a cast operator to a C# array or a `System.Collections.Generic.List<T>`. | ✔ |
| Add support for an `IReadOnlyArray` interface (for you to define) that provides readonly access to the count and each element in the array.  Be sure to appropriately decorate it for covariance and contravariance | ❌ |

**Pull request closed without merging**